Java: Add ConditionalExpr.getBranchExpr(boolean)

Marcono1234 opened the pull request:
Adds the predicate `ConditionalExpr.getBranchExpr(boolean)` (similar to [`ConditionNode.getABranchSuccessor(boolean)`](https://codeql.github.com/codeql-standard-libraries/java/semmle/code/java/ControlFlowGraph.qll/predicate.ControlFlowGraph$ConditionNode$getABranchSuccessor.1.html) and updates current usage of `ConditionalExpr` predicates.

smowton commented on java/ql/src/semmle/code/java/dataflow/Nullness.qll:
```diff
@@ -438,13 +438,9 @@ private predicate varConditionallyNull(SsaExplicitUpdate v, ConditionBlock cond,
   |
-    condexpr.getTrueExpr() = nullExpr() and
+    condexpr.getBranchExpr(branch) = nullExpr() and
     branch = true and
```
`branch` should be free

smowton reviewed (commented):
Considering the are so many uses of `getBranchExpr(_)`, you should probably add `getABranchExpr() = getBranchExpr(_)`

aschackmull commented on java/ql/src/semmle/code/java/controlflow/internal/GuardsLogic.qll:
```diff
@@ -40,19 +40,15 @@ predicate implies_v1(Guard g1, boolean b1, Guard g2, boolean b2) {
-    or
-    cond.getFalseExpr() = boollit and branch = false
+    cond.getBranchExpr(branch) = boollit
   |
```
Style: Replace this `|` with `and` now that the first part is no longer a disjunction.

aschackmull commented on java/ql/src/semmle/code/java/Expr.qll:
```diff
@@ -1213,6 +1206,15 @@ class ConditionalExpr extends Expr, @conditionalexpr {
+   * it is `getFalseExpr()`.
+   */
+  Expr getBranchExpr(boolean branch) {
+    if branch = true then result = getTrueExpr() else result = getFalseExpr()
```
This doesn't bind `branch` so we'll want to replace it with a disjunction of purely positive terms.
```suggestion
    branch = true and result = getTrueExpr() or branch = false and result = getFalseExpr()
```

aschackmull commented on java/ql/src/semmle/code/java/Expr.qll:
```diff
@@ -297,9 +295,7 @@ class CompileTimeConstantExpr extends Expr {
-      if condition = true
-      then result = ce.getTrueExpr().(CompileTimeConstantExpr).getBooleanValue()
-      else result = ce.getFalseExpr().(CompileTimeConstantExpr).getBooleanValue()
+      ce.getBranchExpr(condition).(CompileTimeConstantExpr).getBooleanValue()
```
```suggestion
      result = ce.getBranchExpr(condition).(CompileTimeConstantExpr).getBooleanValue()
```

Marcono1234 commented:
Thanks for your review and sorry for the oversights. I have addressed your review feedback.

(Also feel free to squash the commits of this branch before merging)

aschackmull commented:
The file `ModulusAnalysis.qll` is shared between the Java libraries and the C# libraries, and the CI complains that the copies are out of sync. Could you make the corresponding change to `csharp/ql/src/semmle/code/csharp/dataflow/ModulusAnalysis.qll`? This can usually be done with the convenient helper script `python3 ./config/sync-files.py --latest` (note that this relies on file modification time stamps to select the copy to sync from), but may require a similar interface addition in the C# libs.

aschackmull commented:
```
  Loading file groups from ./config/identical-files.json
  ERROR: Identical files are out of sync
  ERROR: ql/config/sync-files.py:0 - Files from group 'ModulusAnalysis Java/C#' not in sync.
  ERROR: ql/config/sync-files.py:0 - Run this script with a file-name argument among the following to overwrite the remaining files with the contents of that file, or run with the --latest switch to update each group of files from the most recently modified file in the group.
  ERROR: ql/config/sync-files.py:0 -     ./java/ql/src/semmle/code/java/dataflow/ModulusAnalysis.qll
  ERROR: ql/config/sync-files.py:0 -     ./csharp/ql/src/semmle/code/csharp/dataflow/ModulusAnalysis.qll
  FAILED: sync-identical-files.py
```

Marcono1234 commented:
I have synced the changes with C#'s `ModulusAnalysis.qll` and added the predicate `getBranchExpr` to `RangeUtils.qll`. I hope that is what you meant by "may require a similar interface addition in the C# libs", though I am not familiar with how the C# library works.

aschackmull commented:
Perfect, thanks.
